sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamBuiltinAggregations.java

```diff
@@ -554,4 +579,235 @@ public BigDecimal toBigDecimal(BigDecimal record) {
       return record;
     }
   }
+
+  static class CovarAgg implements Serializable {
+    Long count; // number of elements
+    BigDecimal xavg; // average of x elements
+    BigDecimal yavg; // average of y elements
+    BigDecimal covar; // n times the covariance
+
+    public CovarAgg(Long count, BigDecimal xavg, BigDecimal yavg, BigDecimal covar) {
+      this.count = count;
+      this.xavg = xavg;
+      this.yavg = yavg;
+      this.covar = covar;
+    }
+  }
+
+  /**
+   * {@link CombineFn} for <em>Covar_pop and Covar_samp</em> on {@link Number} types.
+   * Covariance Populator and Sample
+   * http://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
+   * <p>
+   * Incremental:
+   *   count -> n
+   *   xavg -> mx_n = mx_(n-1) + [x_n - mx_(n-1)]/n
+   *   yavg -> my_n = my_(n-1) + [y_n - my_(n-1)]/n
+   *   covariance * n -> c_n = c_(n-1) + (x_n - mx_(n-1))*(y_n - my_n)
+   * </p>
+   * <p>
+   * Merge:
+   *   c_X = c_A + c_B + (mx_A - mx_B)*(my_A - my_B)*n_A*n_B/n_X
+   * </p>
+   * <p>
+   *  This one-pass algorithm is stable.
+   * </p>
+   */
+  abstract static class Covar<T extends Number>
```

Comment:
Can the same code be used to calculate both Covariance and Variance? A lot of the logic here seems similar to Variance calculation. Can the fact that Var(x) = Covar(x,x) simplify this code?

Reply:
That's true. I also checked out your refactoring PR #4448 . That's great! I definitely can simplify this code based on your refactoring.